### DISCUSSION TITLE
PyCharm - Cannot open project

### DISCUSSION THREAD

**Discussion by MKPlato** [JetBrains/Qodana#172] [Q&A]
I am using our hosted gitlab CI to publish the report to Qodana cloud. When I click `open file in PyCharm`,  it reports "Cannot find or open the specified project".  From log, I can see it try to open `jetbrains://pycharm/qodana/showMarker?origin=https://gitlab.mygitlab.com/myproject/plato.git&marker_revision=3ac986ca623f30d5d435d16ca5e2e5eccd1b50da&path=src/plato/invitation/services.py:186:129&length=14&message=Local%20variable%20%27inv_identifier%27%20might%20be%20referenced%20before%20assignment`

Qodana is 2023.1 python linter, Pycharm is also latest 2023.1.2.   OS is Mac OSX 13.4. 

Because our hosted gitlab webpage URL and git URL are different, https://gitlab.mygitlab.com/myproject/plato.git and https://git.mygitlab.com/myproject/plato.git, without "lab" in first part of url.  I tried replace it in that URI and open but same result. 

I am not sure what is the indicator to tell PyCharm project name.  Git origin?  or project name in some configuration files under .idea folder?

**Comment by brichbash** [comment]
Hello,

If you have opened your project in PyCharm before, then Qodana should be able to open problem in the IDE. Does your project remote "origin" match the URL from the log?
Also, please share your PyCharm logs after trying to open the report. You can access them by going to **Help | Show Log in Finder**

**Comment by MKPlato** [comment]
My git origin is git@git.mygitlab.com:myproject/plato.git,  but our CI site is git**lab**.mygitlab.com.  However, I tested both origin in that URI, same issue. 

The log doesn't have detail errors:
```
2023-06-05 16:19:36,171 [ 138339]   INFO - #c.i.i.ApplicationLoader - External instance command received
2023-06-05 16:19:36,171 [ 138339]   INFO - #c.i.i.CommandLineProcessor - external URI request:
jetbrains://pycharm/qodana/showMarker?origin=https://gitlab.mygitlab.com:myproject/plato.git&marker_revision=f56fd918f8218ef072f8216e179e20e1773603dd&path=src/plato/invitation/services.py:186:129&length=14&message=Local%20variable%20%27inv_identifier%27%20might%20be%20referenced%20before%20assignment
2023-06-05 16:19:36,789 [ 138957]   INFO - #c.i.w.i.i.j.s.JpsGlobalModelSynchronizerImpl - Saving global entities to files
2023-06-05 16:19:42,522 [ 144690]   INFO - #c.i.w.i.i.j.s.JpsGlobalModelSynchronizerImpl - Saving global entities to files
```

PyCharm error pop:
<img width="502" alt="Screenshot 2023-06-05 at 4 19 39 pm" src="https://github.com/JetBrains/Qodana/assets/95200929/104d270e-304c-4a3a-8426-81cf1e59ba62">

**Comment by MKPlato** [comment]
how does PyCharms locate the project?  I thought it could be by that origin.  I test in various combination.. no luck.  One more guess, my project is a submodule, so it doesn't have `.git` folder in current project folder.  Does PyCharm use that any content from .git folder?

**Reply by brichbash** [reply]
That's correct. We do rely on `.git` in the project root.
Could you try running Qodana from your project root with `.git` setting `--project-dir <relative-path-to-submodule>` flag? This way, Qodana should resolve your VCS data and `qodana.yaml` but only analyze your submodule.

**Comment by MKPlato** [comment]
OK, I updated CI job to root project.  After I manually replaced that origin URI from  gitlab.mygitlab.com to git.mygitlab.com, it works... So it actually still not working for my case because it seems no way to modify that origin URI.  

Furthermore,  open VCS was working if Qodana built from the sub-module project, but failed when it was built on root project, due to Gitlab cannot open a submodule project file from root project URI :( 

If your future development can consider these scenarios, that will be great! Thank you.

**Reply by brichbash** [reply]
You can also try using `--source-dir` instead of `--project-dir` and check if that resolves the issue. The difference between them is that `--project-dir` is used to set a directory that Qodana uses for configuration, while `--source-directory` just limits the analysis scope to the set directory and configures project from the project root.

**Reply by MKPlato** [reply]
This seems not working Open in VCS in submodule project.  Anyway, this is not critical for us.  Just hope in future, it can have a perfect solution for open IDE and VCS by different URL and submodule.

**Comment by MKPlato** [comment]
More information, why our git origin and gitlab CI are using different domain, that is because we are using Cloudflare Zero Trust gateway, we have to use different domain for these two kind of network services.

**Reply by brichbash** [reply]
Could you please try passing to Qodana job an environment variable `QODANA_REMOTE_URL` with your git remote URL via the flag `-e`? This should change the URL from `Open in IDE`

**Reply by MKPlato** [reply]
This QODANA_REMOTE_URL resolved Open file in IDE issue.
